base/iostream.jl

```diff
@@ -336,14 +336,23 @@ function read(s::IOStream, nb::Integer; all::Bool=true)
 end
 
 ## Character streams ##
-const _chtmp = Ref{Char}()
+
 function peekchar(s::IOStream)
-    if ccall(:ios_peekutf8, Cint, (Ptr{Void}, Ptr{Char}), s, _chtmp) < 0
+    chref = Ref{UInt32}()
+    if ccall(:ios_peekutf8, Cint, (Ptr{Void}, Ptr{UInt32}), s, chref) < 0
```

Comment:
current implementation of `ios_peekutf8` also does not check for invalid sequences.

Reply:
Seems a bit wasteful for `ios_peekutf8` to convert the UTF-8 encoding to `UInt32`, and then to convert it back when returning `Char(chref[])`.

Since this seems to be the only function in all of Julia that calls `ios_peekutf8`, we should just re-write `ios_peekutf8` to return `Char`.

Reply:
Agree, this stuff should be optimized to avoid converting back and forth. This PR is WIP and the first step is just to get everything working and all tests passing.

Reply:
This is still in my new PR, which is annoying but I think that all the `ios_` support code needs an overhaul in the 1.x timeframe and ideally we would move most of the buffering code etc. into Julia and it can understand the new `Char` representation and avoid the inefficiency here. Until then I think it's ok to just leave this as it is.